pkg/block/forest.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package block
+
+import (
+	"github.com/oklog/ulid"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+)
+
+// Vertex type implements a vertex of a Forest.
+type Vertex struct {
+	ID       ulid.ULID
+	Value    *metadata.Meta
+	Children []*Vertex
+}
+
+// NewVertex creates a new vertex.
+func NewVertex(id ulid.ULID, value *metadata.Meta) *Vertex {
+	v := &Vertex{
+		ID:       id,
+		Children: []*Vertex{},
+		Value:    value,
+	}
+
+	return v
+}
+
+// Forest type is a collection of trees, where each tree has nodes having relationships.
+type Forest struct {
+	RootVertex *Vertex
+}
+
+// NewForest creates a new forest object.
+func NewForest() *Forest {
+	return &Forest{
+		RootVertex: NewVertex(ulid.MustNew(uint64(0), nil), nil),
+	}
+}
+
+// AddMeta Inserts meta into forest such that the Sources of parent vertex is superset of all child vertices.
+func (f *Forest) AddMeta(id ulid.ULID, metas map[ulid.ULID]*metadata.Meta) bool {
+	return f.addVertex(id, f.RootVertex, metas)
+}
+
+// GetNonRootIDs returns list of ids which are not on root level of forest.
+func (f *Forest) GetNonRootIDs() []ulid.ULID {
+	var ulids []ulid.ULID
+	for _, vertex := range f.RootVertex.Children {
+		ulids = append(ulids, f.iterate(vertex)...)
+		ulids = remove(ulids, vertex.ID)
+	}
+	return ulids
+}
+
+func (f *Forest) iterate(vertex *Vertex) []ulid.ULID {
+	var ulids []ulid.ULID
+	ulids = append(ulids, vertex.ID)
+	for _, childVertex := range vertex.Children {
+		ulids = append(ulids, f.iterate(childVertex)...)
+	}
+	return ulids
+}
+
+func (f *Forest) addVertex(id ulid.ULID, startVertex *Vertex, metas map[ulid.ULID]*metadata.Meta) bool {
+	var rootVertex *Vertex
+	newVertex := NewVertex(id, metas[id])
+	for _, vertex := range startVertex.Children {
+		rootSources := metas[vertex.ID].Compaction.Sources
+		sources := metas[id].Compaction.Sources
+
+		if f.containsSources(rootSources, sources) && f.containsSources(sources, rootSources) { // a block exists with same sources.
```

Comment:
This line is quite long so maybe we can put comment above? And make it a full sentence? (Capital first letter)